### PR TITLE
Update type annotations to the point where mypy is satisfied

### PR DESCRIPTION
**Pull request by iamgreaser**
NOTE: Testingwise I ended up running this and doing a `!join` then a `!quit`, although that was after the first commit and I missed one `from __future__ import annotations` line so I needed another commit. This does touch a lot of files and I'm not sure how you want this PR to be arranged.

If a `botconfig.py` is not provided then running mypy like this (available on pip) should pass:

```
python3 -m mypy .
```

This is good enough to get mypy 0.812 to typecheck things successfully in non-strict mode, although a custom mypy.ini was provided to ignore all failures in `src.messages.*`.

A large chunk of type comments were turned into Python 3.5/3.6 inline annotations.

`from __future__ import annotations` allows Python 3.7+ to not actually attempt to resolve type names when it sees said inline annotations.

Various imports are guarded by the `typing.TYPE_CHECKING` variable so not to introduce potential import loops into the code.

Several assertions were put in places to make mypy happy. In most cases, if the assertions were not there, the code would have crashed anyway.

`UserSet` and `UserList` were confirmed to not need to be generic. `UserDict` on the other hand seemed like the key or the value could be not a `User` and therefore both key and value needed to be generic.

Several cases of `users.User`, mostly those in annotations and type comments, were changed to be just `User`.

`src/settings.py` was for the purposes of this commit deemed to be something longing to be killed with fire and as a result I've largely avoided giving it proper type annotations for some things, especially where it gets extra stuff dumped into it for no reason from `src/wolfgame.py`.

A bug was fixed in `src/status/dying.py` as a result of needing to get the types checked. `None` is falsy, but it was meant to return `True` rather than do a no-value return (which returns `None`). I'm not sure if we want to split this one out or not. (The other case of a no-value return being changed was due to it actually needing to return `None`, so the change is behaviourally-identical.)

### CODE REVIEW

File: src/cats.py [hunk 1]
```diff
@@ -44,7 +46,7 @@
 
 FROZEN = False
 
-ROLES = {}
+ROLES: Dict[str, str] = {}
```
Comment (skizzerz) [inline src/cats.py:49]: This is wrong. You also put `Dict[str, str]` in a handful of other places where it's wrong. Actually check how the type is being used before throwing an arbitrary annotation on it. If you can't figure it out, leave it un-annotated.

File: src/context.py [hunk 2]
```diff
@@ -121,7 +122,9 @@ def wrapper(cls):
 class IRCContext:
     """Base class for channels and users."""
 
-    _messages = defaultdict(list)
+    _messages: Dict[str, List[IRCContext]] = defaultdict(list)
+
+    is_fake: ClassVar[bool] = False
```
Comment (skizzerz) [inline src/context.py:127]: Why is this being added here?

File: src/context.py [hunk 3]
```diff
@@ -318,7 +321,8 @@ def CHANMODES(self) -> Tuple[str, str, str, str]:
         modes = self._features.get("CHANMODES", [])
         while len(modes) < 4:
             modes.append("")
-        return tuple(modes[:4])
+        rA, rB, rC, rD = modes[:4]
+        return (rA, rB, rC, rD)
```
Comment (skizzerz) [inline src/context.py:325]: Why?

File: src/debug/dict.py [hunk 4]
```diff
@@ -46,7 +46,7 @@ def __delitem__(self, k: KT) -> None:
         self._history.add("delitem", k)
         del self._dict[k]
 
-    def __getitem__(self, k: KT) -> VT_co:
+    def __getitem__(self, k: KT) -> VT:
```
Comment (skizzerz) [inline src/debug/dict.py:49]: Why this change? Do you know what covariance is? I'm nearly certain this was correct before, because we may be returning a type that is covariant in VT, not necessarily an instance of VT itself.

File: src/decorators.py [hunk 5]
```diff
@@ -23,22 +25,22 @@
...
 
-_local = _local()
+_local = _local_cls()
 
 # This is a mapping of stringified tracebacks to (link, uuid) tuples
 # That way, we don't have to call in to the website everytime we have
 # another error.
 
-_tracebacks = {}
+_tracebacks: Dict[str, str] = {}
```
Comment (skizzerz) [inline src/decorators.py:43]: Completely wrong type

File: src/decorators.py [hunk 6]
```diff
@@ -192,7 +194,7 @@ def __exit__(self, exc_type, exc_value, tb):
 
 class handle_error:
 
-    def __new__(cls, func=None, *, instance=None):
+    def __new__(cls, func: Optional[Callable[..., None]]=None, *, instance=None):
```
Comment (skizzerz) [inline src/decorators.py:197]: func shouldn't be optional here; it defaulting to None is wrong (should have no default)

File: src/decorators.py [hunk 7]
```diff
@@ -307,6 +309,7 @@ def caller(self, var, wrapper: MessageDispatcher, message: str):
                 return # commands not allowed in alt channels
 
         if "" in self.commands:
+            assert self.func is not None
```
Comment (skizzerz) [inline src/decorators.py:312]: rather than asserting this isn't None everywhere, simply modify the typing of self.func to indicate that it's not Optional

File: src/handler.py [hunk 8]
```diff
@@ -192,8 +194,10 @@ def parse_and_dispatch(var,
                 wrapper.pm(messages["no_force_admin"])
                 return
             if fn.chan:
+                assert channels.Main is not None
```
Comment (skizzerz) [inline src/handler.py:197]: this is never None in the course of normal bot operation (with exception of early bootstrapping), so it should be typed as such rather than adding spurious asserts elsewhere. Ditto users.Bot

File: src/lineparse.py [hunk 9]
```diff
@@ -59,7 +61,8 @@ def add_subparsers(self, **kwargs):
         self.allow_intermixed = False
         return super().add_subparsers(**kwargs)
 
-    def parse_args(self, args: Optional[Sequence[str]] = None, namespace: Optional[Namespace] = None) -> Namespace:
+    def parse_args(self, args: Optional[Sequence[str]] = None, namespace: Optional[Namespace] = None) -> Namespace:  # type: ignore
+        # parse_args has a difficult type signature to follow
```
Comment (skizzerz) [inline src/lineparse.py:65]: Revert this

File: src/messages/formatter.py [hunk 10]
```diff
@@ -103,7 +103,7 @@ def format_field(self, value, format_spec, *, flatten_lists=True):
             value = self._article(value, specs["article"])
             del specs["article"]
         if "!" in specs:
-            from botconfig import CMD_CHAR
+            from botconfig import CMD_CHAR  # type: ignore
```
Comment (skizzerz) [inline src/messages/formatter.py:106]: this is a str

File: src/pregame.py [hunk 11]
```diff
@@ -17,16 +20,19 @@
...
 
 WAIT_LOCK = threading.RLock()
 WAIT_TOKENS = 0
 WAIT_LAST = 0
 
-LAST_START = UserDict() # type: UserDict[users.User, List[datetime, int]]
-LAST_WAIT = UserDict() # type: UserDict[users.User, datetime]
-START_VOTES = UserSet() # type: UserSet[users.User]
-RESTART_TRIES = 0 # type: int
+LAST_START: UserDict[User, List[Union[datetime, int]]] = UserDict()  # FIXME: this should probably be a class containing two fields
```
Comment (skizzerz) [inline src/pregame.py:32]: don't add FIXME comments to existing code without consulting the main devs. We do actually care about and track these.

File: src/roles/bodyguard.py [hunk 12]
```diff
@@ -1,22 +1,26 @@
...
 from src.messages import messages
 from src.status import try_misdirection, try_exchange, add_protection, add_dying
 from src.cats import Wolf
+from src.users import User
 
-GUARDED = UserDict() # type: UserDict[users.User, users.User]
+GUARDED: UserDict[User, User] = UserDict()
 PASSED = UserSet()
-DYING = set()
+DYING: Set[User] = set()
```
Comment (skizzerz) [inline src/roles/bodyguard.py:23]: Should be a UserSet() and additionally cleared in the reset event. Not sure why this wasn't the case already

File: src/roles/matchmaker.py [hunk 13]
```diff
@@ -13,9 +16,12 @@
 from src.status import try_misdirection, try_exchange, add_dying
 from src.cats import Win_Stealer
 
+if TYPE_CHECKING:
+    from src.users import User
+
 MATCHMAKERS = UserSet()
 ACTED = UserSet()
-LOVERS = UserDict()
+LOVERS: UserDict[User, User] = UserDict()
```
Comment (skizzerz) [inline src/roles/matchmaker.py:24]: Wrong type

File: src/roles/warlock.py [hunk 14]
```diff
@@ -13,10 +16,13 @@
...
 
+if TYPE_CHECKING:
+    from src.users import User
+
 register_wolf("warlock")
 
-CURSED = UserDict() # type: UserDict[users.User, users.User]
-PASSED = UserSet() # type: UserSet[users.Set]
+CURSED: UserDict[User, User] = UserDict()
+PASSED: UserSet = UserSet()
```
Comment (skizzerz) [inline src/roles/warlock.py:25]: is the PASSED annotation necessary? Type is already obvious

File: src/settings.py [hunk 15]
```diff
@@ -170,7 +173,7 @@
 
 DOCTOR_IMMUNIZATION_MULTIPLIER = 0.135 # ceil(num_players * multiplier) = number of immunizations
 
-GAME_MODES = {}
+GAME_MODES: Dict[str, Any] = {}
```
Comment (skizzerz) [inline src/settings.py:176]: Wrong type

File: src/status/dying.py [hunk 16]
```diff
@@ -67,7 +67,7 @@ def kill_players(var, *, end_game: bool = True) -> bool:
     with var.GRAVEYARD_LOCK: # FIXME
         if not var.GAME_ID or var.GAME_ID > t:
             #  either game ended, or a new game has started
-            return
+            return True  # CHECK BEFORE PULLING: This used to say just "return" which appears to have been an actual bug. --GM
```
Comment (skizzerz) [inline src/status/dying.py:70]: remove comment

File: src/status/forcevote.py [hunk 17]
```diff
@@ -16,30 +19,30 @@
 # FORCED_TARGETS is used to see which targets are being force-voted for.
 # If multiple calls to force_vote() force different targets, the union of all of those targets is taken.
 
-FORCED_COUNTS = UserDict() # type: UserDict[users.User, int]
-FORCED_TARGETS = UserDict() # type: UserDict[users.User, UserSet]
+FORCED_COUNTS: UserDict[User, int] = UserDict()
+FORCED_TARGETS: UserDict[User, UserSet] = UserDict()
 
-def _add_count(var, votee : users.User, amount : users.User) -> None:
+def _add_count(var, votee : User, amount : int) -> None:
```
Comment (skizzerz) [inline src/status/forcevote.py:25]: bad annotation format, no space before colons (yes I know it was that way before, but you can fix it since you're already touching those lines)

File: src/status/silence.py [hunk 18]
```diff
@@ -1,10 +1,14 @@
...
+from typing import TYPE_CHECKING
+
 from src.containers import UserSet
 from src.decorators import event_listener
 from src.messages import messages
 
 __all__ = ["add_silent", "is_silent"]
 
-SILENT = UserSet() # type: UserSet[users.User]
+SILENT: UserSet = UserSet()
```
Comment (skizzerz) [inline src/status/silence.py:11]: Again, is this necessary?

File: src/users.py [hunk 19]
```diff
@@ -10,16 +12,16 @@
...
 from src.match import Match
 
-import botconfig
+import botconfig  # type: ignore
 
 __all__ = ["Bot", "predicate", "get", "add", "users", "disconnected", "complete_match",
            "parse_rawnick", "parse_rawnick_as_dict", "User", "FakeUser", "BotUser"]
 
-Bot = None # bot instance
+Bot: Optional[BotUser] = None # bot instance
```
Comment (skizzerz) [inline src/users.py:20]: Will only be none during bootstrap; main code should treat this as non-Optional

File: src/wolfgame.py [hunk 20]
```diff
@@ -72,62 +74,65 @@
...
 # Game Logic Begins:
 
-var.LAST_STATS = None
-var.LAST_ADMINS = None
-var.LAST_GSTATS = None
-var.LAST_PSTATS = None
-var.LAST_RSTATS = None
-var.LAST_TIME = None
-var.LAST_GOAT = UserDict() # type: UserDict[users.User, datetime]
+var.LAST_STATS = None  # type: ignore
```
Comment (skizzerz) [inline src/wolfgame.py:86]: omit all ignores. If mypy is throwing an error for these lines, suppress the individual error instead of opting out of the type system entirely. If you don't know the type, omit the annotation. Otherwise fill in the correct type.

**Review by skizzerz — CHANGES_REQUESTED** [review]
Thanks for your effort, but a lot of this is quite sloppy. As-is this is unmergeable.

I've asked for clarification in a number of areas that I wasn't personally very sure about, please give a solid explanation for your changes there or revert them.

File: src/wolfgame.py [hunk 21]
```diff
@@ -953,6 +958,7 @@ def fjoin(var, wrapper: MessageDispatcher, message: str):
 
     parts = re.split(" +", message)
     to_join = []
+    assert isinstance(wrapper.target, Channel)
```
Comment (Vgr255) [inline src/wolfgame.py:961]: Why is this necessary?
